File: drone.c
#include "drone.h"
#include<stdlib.h>
#define _USE_MATH_DEFINES
#include <math.h>
#ifndef M_PI
#define M_PI (3.14159265358979323846)
#endif
#include "vec2.h"

static uint32_t ids =0;

const Obstacle compute_obstacle(Drone *d1, Drone* d2)
{
	// Minkowski addition
	double r = d1->size + d2->size;
	vec2 center;
	//Computing tangent lines to circle passing through the point self.position
	double dx = center.x - d2->position.x;
	double a = dx * dx - r * r;
	double b = 2 * dx * (center.y - d2->position.y);
	double c = (d2->position.y - center.y) * (d2->position.y - center.y) - r * r;
	double Delta = b * b - 4 * a * c;

	//Angular coefficient
	double m1 = (-b + sqrt(Delta)) / (2 * a);
	double m2 = (-b - sqrt(Delta)) / (2 * a);
	//Intersection with y axis
	double q1 = center.y - m1 * center.x;
	double q2 = center.y - m2 * center.x;

	//(xt1,yt1) - first tangent point.
	double a1 = 1 + m1 * m1;
	double b1 = 2 * m1 * q1 - 2 * d2->position.x - m1 * 2 * d2->position.y;

	double xt1 = (-b1) / (2 * a1);
	double yt1 = m1 * xt1 + q1;

	//(xt2,yt2) - Second tangent point
	double a2 = 1 + m2 * m2;
	double b2 = 2 * m2 * q2 - 2 * d2->position.x - m2 * 2 * d2->position.y;

	double xt2 = (-b2) / (2 * a2);
	double yt2 = m2 * xt2 + q2;

	//Construct obstacle
	Obstacle o;
	o.position = d2->position;
	o.radius = r;
	o.T1.x = xt1;
  	o.T1.y = yt1;
	o.T2.x = xt2;
  	o.T2.y = yt2;

	return o;
}

barycoords barycentric(vec2 A, vec2 B, vec2 C, vec2 P){
  	barycoords bc;
  	bc.gamma = ((A.y - B.y) * P.x + (B.x - A.x) * P.y + A.x * B.y - B.x * A.y) /
				((A.y - B.y) * C.x + (B.x - A.x) * C.y + A.x * B.y - B.x * A.y);
	bc.beta = ((A.y - C.y) * P.x + (C.x - A.x) * P.y + A.x * C.y - C.x * A.y) /
			   ((A.y - C.y) * B.x + (C.x - A.x) * B.y + A.x * C.y - C.x * A.y);
	bc.alpha = 1 - bc.beta - bc.gamma;
  	return bc;

}
    
Drone DR_newDrone(double x, double y,double vx,double vy, double size){
	Drone d;
	d.id = ids;
	ids++;
	d.position.x = x;
	d.position.y = y;
	d.speed.x = vx;
	d.speed.y = vy;
	d.waypoints = malloc(2*sizeof(vec2));
	d.wp_len = 2;
	d.curr_wp = 0;
	d.size = size;
	return d;
}

void DR_move(Drone* d, double dt){
	if(d->curr_wp < 0){
		return;
	}
	if(v2_distance(&d->position,&d->waypoints[d->curr_wp]) < d->size){
		DR_pop_waypoint(d);
	}
	
	DR_goto(d,d->waypoints[d->curr_wp]);
	vec2 dP = v2_prodK(&d->speed,dt);
	d->position = v2_add(&d->position,&dP);

}

void DR_goto(Drone* d, vec2 waypoint){
  
  vec2 dir = v2_norm(&d->speed);
  
  vec2 dirp = v2_sub(&waypoint, &d->position);
  
  dirp = v2_norm(&dirp);
  
  double C = v2_dot(&dir,&dirp);
  
  if(C > 1.0){
    C=1.0;
  }
  else if(C < -1.0){
    C=-1.0;
  }

  double angle = -acos(C);
  //steer
  
  d->speed = v2_rotate(&d->speed,angle);

}

bool DR_collision(Drone* d1, Drone* d2){
	Obstacle o = compute_obstacle(d1,d2);
	vec2 dif = v2_sub(&d1->speed,&d2->speed);
	vec2 ds = v2_add(&dif,&d1->position);
	barycoords bc = barycentric(o.position,o.T1,o.T2,ds);
	if(bc.alpha>0 && bc.beta >0 && bc.gamma>0){
		return true;
	}

    return false;

}
void DR_avoid(Drone* d, Drone* d2){
	if(DR_collision(d,d2)){
		vec2 escape = v2_rotate(&d->speed,M_PI/2);
		escape = v2_norm(&escape);
		escape = v2_prodK(&escape,2*(d->size+d2->size));
		escape = v2_add(&escape,&d->position);
		DR_push_waypoint(d,escape);
		
	}
}
void DR_push_waypoint(Drone* d, vec2 wp){
	
	if(d->curr_wp==(d->wp_len-1)){
		d->wp_len = (d->wp_len+(d->wp_len>>1));
		d->waypoints = realloc(d->waypoints,(d->wp_len+(d->wp_len>>1))*sizeof(vec2));	 
	}
	d->curr_wp += 1;
	d->waypoints[d->curr_wp] = wp;
}
void DR_pop_waypoint(Drone* d){
	d->curr_wp -= 1;
}

void DR_freeDrone(Drone*d){
	free(d->waypoints);
}

File: prototypes/drone.py
from vector2 import *
from dataclasses import dataclass
from collections import deque
from math import sqrt,cos,sin,acos

@dataclass
class Obstacle:
    size :float
    pos :vec2
    T1 :vec2
    T2 :vec2

#kinematic model of the drone
def barycentric( P1,  P2,  P3,  P):
    

    dg = ((P1.y - P2.y) * P3.x + (P2.x - P1.x) * P3.y + P1.x * P2.y - P2.x * P1.y)
    db = ((P1.y - P3.y) * P2.x + (P3.x - P1.x) * P2.y + P1.x * P3.y - P3.x * P1.y)

    if db == 0 or dg == 0:
        return (-1,-1,-1)
    gamma = ((P1.y - P2.y) * P.x + (P2.x - P1.x) * P.y + P1.x * P2.y - P2.x * P1.y) / dg
        
    beta = ((P1.y - P3.y) * P.x + (P3.x - P1.x) * P.y + P1.x * P3.y - P3.x * P1.y) / db
        
    alpha = 1 - beta - gamma
    return alpha, beta, gamma

class Drone:
    def __init__(self, id, x, y, vx, vy, size) -> None:
        self.id = id
        self.pos = vec2(x,y)
        self.vel = vec2(vx,vy)
        self.size = size
        self.trajectory = deque()
    
    def move(self,dt):
        self.pos += self.vel*dt

    def accelerate(self,dv,dt):
        self.vel += dv*dt

    def stop(self):
        self.vel.x = 0
        self.vel.y = 0

    def start(self,vx,vy):
        self.vel = vec2(vx,vy)    
    
    def collide(self,drone):
        obs = self.obstacle(drone)
        # (V1 - V2) + P (The translation is needed to check if the difference falls into the triangle)
        DV = self.vel - drone.vel + self.pos
        # computes barycentric coordinates
        a,b,g = barycentric(obs.T1, obs.T2, self.pos, DV.x, DV.y)
        return (a>0 and b>0 and g>0)

    def compute_trajectory(self,P, steps=100):
        
        Pm1 = self.vel*self.size
        try:
            pstart = self.trajectory[-1]
        except:
            pstart = self.pos

        for i in range(1,steps+1): #11 steps just because...
            self.trajectory.append(spline(pstart,Pm1,P,i/steps))
        
    def obstacle(self, d2):
        # Minkowski addition
        r = self.size+d2.size

        # Computing tangent lines to circle passing through the point self.pos
        dx = self.pos.x-d2.pos.x

        a = dx**2 - r**2
        b = 2*dx*(d2.pos.y-self.pos.y)
        c = (d2.pos.y-self.pos.y)**2 - r**2
        Delta = b**2 - 4*a*c
        if Delta < 0:
            Delta = 0
        # Angular coefficient
        m1 = (-b + sqrt(Delta))/(2*a)
        m2 = (-b - sqrt(Delta))/(2*a)
       
        # Intersection with y axis
        q1 = self.pos.y-m1*self.pos.x
        q2 = self.pos.y-m2*self.pos.x

        # P1 - first tangent point.
        a1 = 1+m1**2
        b1 = 2*m1*q1-2*d2.pos.x-m1*2*d2.pos.y

        xt1 = (-b1)/(2*a1)
        yt1 = m1*xt1+q1

        # P2 - second tangent point.
        a2 = 1+m2**2
        b2 = 2*m2*q2-2*d2.pos.x-m2*2*d2.pos.y

        xt2 = (-b2)/(2*a2)
        yt2 = m2*xt2+q2
        return Obstacle(r, d2.pos,vec2(xt1,yt1),vec2(xt2,yt2))
        
    def steer_towards(self,P):
        M = self.vel.mod()
        dirp = (P-self.pos).norm()
        
        self.vel = dirp*M

